src/firejail/template.c

```diff
@@ -0,0 +1,429 @@
+			if (str[i] == '.' && str[i+1] == '.')
+				return -EINVAL;
+
+			break;
+		}
+	}
+
+	return 0;
+}
+
+void check_template(const char *arg) {
+	Template *tmpl;
+	const char *key;
+	const char *value;
+	char *delim = TEMPLATE_KEY_VALUE_DELIM;
+	char *macro = NULL;
+	char *macro_value = NULL;
+	int err = 0;
+
+	fprintf(stdout, "Check %s list %p\n", arg, tmpl_list);
+
+	/* Only alphanumeric chars in template key. */
+	key = strtok((char*)arg, delim);
+	if (is_valid_str(key, STR_CHECK_ALNUM) == -EINVAL) {
+		fprintf(stderr, "Error invalid template key \"%s\"", key);
+		exit(1);
+	}
+
+	/* Only a-zA-Z0-9_ */
+	value = strtok(NULL, delim);
+	if (is_valid_str(value, STR_CHECK_DBUS) == -EINVAL) {
+		fprintf(stderr, "Error invalid template value in \"%s:%s\"",
+								key, value);
+		exit(1);
+	}
+
+	if (asprintf(&macro, "${%s}", key) == -1)
+		errExit("asprintf");
+
+	/* Hardcoded macro or XDG value is not allowed to be overridden. */
+	macro_value = resolve_macro(macro);
+	if (macro_value) {
+		fprintf(stderr, "Error override of \"${%s}\" is not permitted",
+									key);
+		err = -EPERM;
+		free(macro_value);
+		goto out;
+	}
+
+	tmpl = template_new(key, value);
+	err = template_add(tmpl);
+
+	switch (err) {
+	case 0:
+		goto out;
+	case -EINVAL:
+		fprintf(stderr, "Error invalid template key \"%s\" "
+						"value \"%s\"\n", key, value);
+		break;
+	case -EALREADY:
+		fprintf(stderr, "Error template key \"%s\" already exists\n",
+								key);
+		break;
+	}
+
+	template_free(tmpl);
+
+out:
+	free(macro);
+
+	if (err)
+		exit(-err);
+}
+
+static char* append_to_string(char *str1, const char *str2)
+{
+	size_t len;
+
+	if (!str2)
+		return str1;
+
+	if (!str1)
+		return strdup(str2);
+
+	len = strlen(str2) + 1;
+	str1 = (char*)realloc(str1, strlen(str1) + len);
+
+	return strncat(str1, str2, len);
+}
+
+const char *internal_keys[] = { "HOME", "CFG", "RUNUSER", "PATH", "PRIVILEGED",
+				NULL };
+
+static int is_internal_macro(const char *key)
+{
+	char *macro;
+	char *macro_value;
+	int err = 0;
+	int i;
+
+	for (i = 0; internal_keys[i]; i++) {
+		if (!strncmp(key, internal_keys[i], strlen(internal_keys[i]))) {
+			err = -EEXIST;
+			goto exit;
+		}
+	}
+
+	if (asprintf(&macro, "${%s}", key) == -1)
+		errExit("asprintf");
+
+	macro_value = resolve_macro(key);
+	if (macro_value) {
+		err = -EEXIST;
+		free(macro_value);
+	}
+
+	free(macro);
+exit:
+	return err;
+}
+
+static char *process_key_value(char *container, char *key, int *err)
+{
+	char *str;
+	char *token;
+	char *saveptr;
+	const char *delim = TEMPLATE_KEY_MACRO_SUB_DELIMS;
+	const char *value;
+
+	for (str = key; ; str = NULL) {
+		token = strtok_r(str, delim, &saveptr);
+		if (!token)
+			break;
+
+		if (is_internal_macro(token) == -EEXIST) {
+			char *macro;
+
+			if (asprintf(&macro, "${%s}", token) == -1)
+				errExit("asprintf");
+
+			container = append_to_string(container, macro);
+			free(macro);
+			continue;
+		}
+
+		/* Only the first iteration is the template */
```

Comment:
I think I lost the plot completely.

Reply:
Well since this tokenizes first by `$` there can be cases of: `include /path/to/${template1}/something/${template2}/somewhere` ->

1: tokens are a:`/path/to`,  b:`{template1}something/` and c:`{template2}/somewhere`
2: pass only a and b to this functon
3: Only the first item, in b:`template1 and c:`template2` are considered proper keys, the remains is just added to string.

I have an external test app for checking these cases. So in general when parsing the whole string in no circumstances the string should start in profile/permission file with a template.

Reply:
Ok, thanks for the clarification (this is "second level slicer" and you probably meant "2: pass only b and c to this function"?).

If needed, how does one have literal '$' / '{' chars in the final product? I have a feeling that eventually this kind of simple slicing loop will fail to handle something and becomes confusingly complex with special case handling or needs to be replaced with a bit more holistic decode+tokenize logic that allows escaping etc.

Reply:
Yes, `b` and `c` in that 2. My bad.

I have the `process_key_value()` implemented in a way that if there is an internal macro it will have its `${INTERNAL_MACRO}` format in the string. I'm eager to fix that macro.c to have simpler checker but lets see..

Reply:
And I'm not sure how complex this really needs to be. But anyways I haven't worked with Firejail for long so there may be something I've missed but from my point of view the lines in the profiles/permissions do seem to be quite simple.